File: lib/market.js
'use strict';

// Geometric Brownian Motion price engine with correlated sector moves

class MarketEngine {
  constructor(symbols, rng) {
    this.rng = rng;
    this.symbols = symbols;
    this.state = new Map();

    // Initialize per-symbol state
    for (const sym of symbols) {
      this.state.set(sym.symbol, {
        price: sym.basePrice,
        tickSize: sym.tickSize,
        volatility: sym.volatility,
        sector: sym.sector,
        volume: 0,
      });
    }

    // Base params
    this.baseDailyVol = 0.02;    // ~2% daily volatility
    this.sectorCorrelation = 0.6; // intra-sector correlation
    this.ticksPerDay = 23400;     // 6.5 hours * 3600
    this.spikeProb = 0.02;        // 2% chance of volume spike
    this.baseVolume = 100;
  }

  // Advance one tick for all symbols, returns array of tick messages
  tick() {
    // Generate one sector shock per sector
    const sectorShocks = new Map();
    const sectors = [...new Set(this.symbols.map(s => s.sector))];
    for (const sec of sectors) {
      sectorShocks.set(sec, this.rng.gaussian(0, 1));
    }

    const ticks = [];

    for (const sym of this.symbols) {
      const st = this.state.get(sym.symbol);
      const perTickVol = (this.baseDailyVol * st.volatility) / Math.sqrt(this.ticksPerDay);

      // Correlated return: sector component + idiosyncratic component
      const sectorZ = sectorShocks.get(st.sector);
      const idioZ = this.rng.gaussian(0, 1);
      const z = this.sectorCorrelation * sectorZ +
                Math.sqrt(1 - this.sectorCorrelation * this.sectorCorrelation) * idioZ;

      // GBM step: S(t+1) = S(t) * exp((mu - sigma^2/2)*dt + sigma*sqrt(dt)*Z)
      const drift = -0.5 * perTickVol * perTickVol; // risk-neutral
      const logReturn = drift + perTickVol * z;
      let newPrice = st.price * Math.exp(logReturn);

      // Snap to tick size
      newPrice = Math.round(newPrice / st.tickSize) * st.tickSize;
      newPrice = Math.max(st.tickSize, newPrice); // floor at one tick

      st.price = newPrice;

      // Volume
      let vol = Math.round(this.baseVolume * (0.5 + this.rng.next()));
      if (this.rng.chance(this.spikeProb)) {
        vol = Math.round(vol * (3 + this.rng.next() * 7)); // 3x-10x spike
      }
      st.volume += vol;

      // Build spread
      const halfSpread = st.tickSize * (1 + Math.floor(this.rng.next() * 3));
      const bid = Math.round((newPrice - halfSpread) / st.tickSize) * st.tickSize;
      const ask = Math.round((newPrice + halfSpread) / st.tickSize) * st.tickSize;

      ticks.push({
        symbol: sym.symbol,
        lastPrice: +newPrice.toFixed(4),
        volume: vol,
        bid: +Math.max(st.tickSize, bid).toFixed(4),
        ask: +ask.toFixed(4),
      });
    }

    return ticks;
  }

  // Get current price for a symbol
  getPrice(symbol) {
    const st = this.state.get(symbol);
    return st ? st.price : 0;
  }
}

module.exports = { MarketEngine };

File: go-feed/internal/engine/market.go
package engine

import (
	"math"
	"sync"

	"github.com/ndrandal/feed-simulator/go-feed/internal/symbol"
)

const (
	baseDailyVol    = 0.02  // 2% daily volatility
	sectorBlend     = 0.60  // 60% sector shock, 40% idiosyncratic
	driftPerTick    = 0.0   // zero drift for simulation
	ticksPerDay     = 86400 // approximate, for vol scaling
)

// MarketEngine drives GBM price movement with sector-correlated returns.
type MarketEngine struct {
	mu     sync.RWMutex
	rng    *RNG
	prices map[uint16]float64   // locate -> current price
	syms   []symbol.Symbol
	byLoc  map[uint16]*symbol.Symbol

	// sector shocks generated once per tick cycle
	sectorShocks map[symbol.Sector]float64
}

// NewMarketEngine creates a price engine for all symbols.
func NewMarketEngine(rng *RNG, syms []symbol.Symbol) *MarketEngine {
	prices := make(map[uint16]float64, len(syms))
	byLoc := make(map[uint16]*symbol.Symbol, len(syms))
	for i := range syms {
		prices[syms[i].LocateCode] = syms[i].BasePrice
		byLoc[syms[i].LocateCode] = &syms[i]
	}
	return &MarketEngine{
		rng:          rng,
		prices:       prices,
		syms:         syms,
		byLoc:        byLoc,
		sectorShocks: make(map[symbol.Sector]float64),
	}
}

// GenerateSectorShocks produces one gaussian shock per sector.
// Call this once per tick cycle before ticking individual symbols.
func (m *MarketEngine) GenerateSectorShocks() {
	m.mu.Lock()
	defer m.mu.Unlock()
	for _, sec := range symbol.Sectors() {
		m.sectorShocks[sec] = m.rng.Gaussian()
	}
}

// Tick advances the price for a single symbol and returns the new price.
// GBM: S(t+1) = S(t) * exp(drift + vol * Z)
func (m *MarketEngine) Tick(locateCode uint16) float64 {
	m.mu.Lock()
	defer m.mu.Unlock()

	sym := m.byLoc[locateCode]
	if sym == nil {
		return 0
	}

	price := m.prices[locateCode]

	// Per-tick volatility: daily vol / sqrt(ticks_per_day) * symbol multiplier
	tickVol := baseDailyVol / math.Sqrt(ticksPerDay) * sym.VolatilityMultiplier

	// Blended shock: sector + idiosyncratic
	sectorZ := m.sectorShocks[sym.Sector]
	idioZ := m.rng.Gaussian()
	z := sectorBlend*sectorZ + (1-sectorBlend)*idioZ

	// GBM step
	logReturn := driftPerTick + tickVol*z
	price *= math.Exp(logReturn)

	// Snap to tick size, floor at 1 tick
	price = math.Round(price/sym.TickSize) * sym.TickSize
	if price < sym.TickSize {
		price = sym.TickSize
	}

	m.prices[locateCode] = price
	return price
}

// Price returns the current price for a symbol.
func (m *MarketEngine) Price(locateCode uint16) float64 {
	m.mu.RLock()
	defer m.mu.RUnlock()
	return m.prices[locateCode]
}

// SetPrice sets the price for a symbol (used when restoring from DB).
func (m *MarketEngine) SetPrice(locateCode uint16, price float64) {
	m.mu.Lock()
	defer m.mu.Unlock()
	m.prices[locateCode] = price
}

// AllPrices returns a snapshot of all current prices.
func (m *MarketEngine) AllPrices() map[uint16]float64 {
	m.mu.RLock()
	defer m.mu.RUnlock()
	out := make(map[uint16]float64, len(m.prices))
	for k, v := range m.prices {
		out[k] = v
	}
	return out
}
